File: CTCI/src/TreesGraphs/Successor.java
package TreesGraphs;

/**
 * Created by rishabh on 30/07/17.
 */
public class Successor {

    public TreeNode leftMostChild(TreeNode node){
        TreeNode head = node;
        while(head.left != null){
            head = head.left;
        }
        return head;
    }

    public TreeNode successor(TreeNode node){
        if(node.right != null){
            return leftMostChild(node.right);
        } else {
            TreeNode q = node.parent;
            TreeNode x = node;
            while(x != null && x.left != q){
                q = x;
                x = x.parent;
            }
            return x;
        }
    }
}

File: CTCI/trees/successor.py
class Successor:

	def inOrderSucc(self, n):
		if(n == None): return None

		if(n.right != None):
			return self.leftMostChild(n.right)
		else:
			TreeNode q = n
			TreeNode parent = q.parent

			while(parent != None and parent.left != q):
				q = parent
				parent = parent.parent
			return parent

	def leftMostChild(self, n):
		if(n == None): return None
		else:
			while(n.left !=None):
				n = n.left
			return n